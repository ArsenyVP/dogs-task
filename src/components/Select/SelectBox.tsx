import React from 'react'
import { FormControl, NativeSelect } from '@material-ui/core'

interface SelectBoxProps {
    value: string | unknown,
    handleChange: (e: React.ChangeEvent<{ name?: string; value: unknown }>) => void,
    arrayValues: any[] | undefined,
    requiredValidation: boolean
}

const SelectBox: React.FC<SelectBoxProps> = ({ value, handleChange, arrayValues, requiredValidation }: SelectBoxProps) => {
    return (
        <FormControl required={requiredValidation}>
            <NativeSelect
                value={value}
                onChange={handleChange}
                name="breed"
            >
                <option aria-label="None" value="" />
                {
                    arrayValues && arrayValues.map((br: string | number, index: number) => {
                        return <option key={index}>{br}</option>
                    })
                }
            </NativeSelect>
        </FormControl>
    )
}

export default SelectBox
