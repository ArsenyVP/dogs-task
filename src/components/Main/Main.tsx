import React, { useState, useEffect, useCallback } from 'react';
import { Container, Typography, Box, Button, CircularProgress } from '@material-ui/core';
import SelectBox from '../Select/SelectBox';
import PictureList from '../PictureList/PictureList';
import { useStyles } from '../../styles';

const Main: React.FC = () => {
    const classes = useStyles();

    const [dogBreeds, setDogsBreed] = useState<string[] | undefined>(undefined);
    const [breed, setBreed] = useState<string | unknown>('');
    const [breedImages, setBreedImages] = useState<string[] | undefined>(undefined);

    const [subBreeds, setSubBreeds] = useState<string[] | undefined>(undefined);
    const [subBreedInfo, setSubBreedInfo] = useState<any>("");

    const [numbers, setNumbers] = useState<any>(undefined)
    const numbersArray = Array.from(Array(51).keys());
    const [isLoading, setIsLoading] = useState<boolean>(false);

    const handleChangeBreed = useCallback(
        (e: React.ChangeEvent<{ name?: string; value: unknown }>) => {
            setBreed(e.target.value);
            setSubBreedInfo(undefined)
            setSubBreeds(undefined)
        }, [setBreed, setSubBreedInfo, setSubBreeds]
    )

    const handleChangeSubBreed = useCallback(
        (e: React.ChangeEvent<{ name?: string; value: unknown }>) => {
            setSubBreedInfo(e.target.value);
        }, [setSubBreedInfo])

    const handleChangeNumber = useCallback(
        (e: React.ChangeEvent<{ name?: string; value: unknown }>) => {
            setNumbers(e.target.value);
        }, [setNumbers]
    )

    const fetchBreeds = () => {
        fetch("https://dog.ceo/api/breeds/list/all").then(res => {
            return res.json()
        }).then(res => {
            const breeds = res.message;
            const result = [];
            for (let br in breeds) {
                result.push(br)
            }
            setDogsBreed(result);
        })
    }

    const fetchSubBreed = useCallback(
        () => {
            if (breed) {
                fetch(`https://dog.ceo/api/breed/${breed}/list`).then(res => {
                    return res.json();
                }).then(res => {
                    if (res.message.length > 0) {
                        setSubBreeds(res.message);
                    }
                })
            }
        }, [breed, setSubBreeds]
    )

    const fetchBreedImages = () => {
        setIsLoading(true)
        if (breed) {
            fetch(`https://dog.ceo/api/breed/${breed}/${subBreedInfo ? subBreedInfo + "/" : ""}images/random/${numbers}`).then(res => {
                return res.json();
            }).then(res => {
                setBreedImages(res.message)
                setIsLoading(false);
            })
        }
    }

    const getRandomImageByBreed = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        fetchBreedImages();
    }

    useEffect(() => {
        fetchBreeds();
    }, [])

    useEffect(() => {
        fetchSubBreed();
    }, [breed, fetchSubBreed])


    return (
        <Container>
            <Typography className={classes.title} variant={"h3"} component={"h1"}>Dogs Task</Typography>

            <Box className={classes.containerInputs}>
                <form className={classes.formContainer} onSubmit={getRandomImageByBreed}>
                    <Box className={classes.fieldBox}>
                        <SelectBox requiredValidation={true} value={breed} handleChange={handleChangeBreed} arrayValues={dogBreeds} />
                    </Box>
                    <Box className={classes.fieldBox}>
                        {
                            subBreeds && <SelectBox requiredValidation={false} value={subBreedInfo} handleChange={handleChangeSubBreed} arrayValues={subBreeds} />
                        }
                    </Box>
                    <Box className={classes.fieldBox}>
                        <SelectBox requiredValidation={true} value={numbers} handleChange={handleChangeNumber} arrayValues={numbersArray} />
                    </Box>
                    <Box>
                        <Button type="submit" variant="contained" color="primary">View Images</Button>
                    </Box>
                </form>
            </Box>
            <Box className={classes.containerImages}>
                {
                    isLoading ? <CircularProgress /> :
                        <PictureList breedImages={breedImages} classes={classes} />
                }
            </Box>
        </Container>
    )
}

export default Main
