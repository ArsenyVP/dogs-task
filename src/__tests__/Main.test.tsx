import React from 'react';
import { shallow } from 'enzyme';
import App from '../App';
import Main from '../components/Main/Main';
import SelectBox from '../components/Select/SelectBox';

describe("Main Component", () => {
    it("App contains Main Component", () => {
        const AppComponent = shallow(<App />);

        expect(AppComponent.contains(<Main />)).toBeTruthy();
    })

    it("SelectBox has props's value", () => {
        const SelectComponent = shallow(<SelectBox
            value="breed"
            handleChange={(e) => e.preventDefault()}
            arrayValues={["array"]}
            requiredValidation={false}
        />
        )

        expect(SelectComponent.props()).toBeTruthy()
    })
})