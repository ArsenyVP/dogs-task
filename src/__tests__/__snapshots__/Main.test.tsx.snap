// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Component Test works 1`] = `<Main />`;
